src/Fl_SVG_Image.cxx

```diff
@@ -90,143 +144,165 @@ float Fl_SVG_Image::svg_scaling_(int W, int H) {
-    }
-
-#else // ! HAVE_LIBZ
-    // without libz, read .svg file
-    FILE *fp = fl_fopen(filename, "rb");
-    if (fp) {
-      fseek(fp, 0, SEEK_END);
-      long size = ftell(fp);
-      fseek(fp, 0, SEEK_SET);
-      filedata = (char*)malloc(size+1);
-      if (filedata) {
-        if (fread(filedata, 1, size, fp) == size) {
-          filedata[size] = '\0';
+  // yes, this is a const cast to avoid duplicating user supplied data
+  uchar *data = (uchar*)in_data;
+
+  // this is to make it clear what we are doing
+  const char *sharedname = data ? name : NULL;
+  const char *filename = data ? NULL : name;
+
+  // prepare with error data, so we can just return if an error occurs
+  d(-1);
+  ld(ERR_FORMAT);
+
+  // if we are reading from a file, just read the entire file into a memory block
+  if (!data) {
+    FILE *f = fl_fopen(filename, "rb");
+    if (f) {
+      fseek(f, 0, SEEK_END);
+      length = ftell(f);
+      fseek(f, 0, SEEK_SET);
+      data = (uchar*)malloc(length+1);
+      if (data) {
+        if (fread((void*)data, 1, length, f) == length) {
+          data[length] = 0;
         } else {
-          free(filedata);
-          filedata = NULL;
+          free((void*)data);
+          data = NULL;
         }
       }
-      fclose(fp);
+      fclose(f);
     }
-#endif // HAVE_LIBZ
-    if (!filedata) ld(ERR_FILE_ACCESS);
-  } else { // handle non-gzip'ed svg data as a char string
-    // XXX: Make internal copy -- nsvgParse() modifies filedata during parsing (!)
-    filedata = in_filedata ? fl_strdup((const char *)in_filedata) : NULL;
+    if (!data) return;
   }
 
-  // filedata is NULL or contains SVG data as a char string
-  if (filedata) {
-    counted_svg_image_->svg_image = nsvgParse(filedata, "px", 96);
-    free(filedata);     // made with svg_inflate|malloc|strdup
-    if (counted_svg_image_->svg_image->width == 0 || counted_svg_image_->svg_image->height == 0) {
-      d(-1);
-      ld(ERR_FORMAT);
-    } else {
-      w(int(counted_svg_image_->svg_image->width + 0.5));
-      h(int(counted_svg_image_->svg_image->height + 0.5));
+  // now if our data is compressed, we use zlib to infalte it
+#if defined(HAVE_LIBZ)
+  if (length==0 || length>10) {
+    if (data[0] == 0x1f && data[1] == 0x8b) {
+      // this is gzip compressed data, so we decompress it and preplace the data array
+      uchar *uncompressed_data = NULL;
+      size_t uncompressed_data_length = 0;
+      int err = svg_inflate(data, length, uncompressed_data, uncompressed_data_length);
+      if (err == Z_OK) {
+        // replace compressed data with uncompressed data
+        if (in_data == NULL) free(data);
+        length = (size_t)uncompressed_data_length;
+        data = (uchar*)uncompressed_data;
+        data[length] = 0;
+      }
     }
-  } else if (copy_source) {
-    w(copy_source->w());
-    h(copy_source->h());
+  }
+#endif // HAVE_LIBZ
+
+  // now our SVG data should be in text format in `data`, terminated by a NUL
```

Comment:
Sorry for this: but what if not? What if HAVE_LIBZ is not defined? It **should** always be defined, but ...

Reply:
In line 250, `ld` is preset with `ERR_FORMAT`. If we deliver compressed data without zlib present, `nsvgParse()` will error out and `ERR_FORMAT` will be returned.

Reply:
I did add some error handling for compressed data and no zlib. It's much safer.